Add initial dashboard network protocol definition file

drewnoakes opened the pull request:
This `.proto` file defines a contract between the server `Aspire.Hosting` and client `Aspire.Dashboard`.

It supports:

- Getting basic information about the server (name and version).
- Subscribing to resources, receiving both an initial snapshot and a stream of updates as they happen.
- ~A heartbeat mechanism, to detect when the connection has dropped and allow prompting the user and/or reconnecting.~
- Support for arbitrary kinds of resources. Not limited to known types (project, container, executable).
- Support for arbitrary commands on resources, intended for actions such as start/stop/refresh/restart/etc. The server defines the commands available for each resource, and the UI can show them.

It's added to both the server and client projects, and code generation is enabled.

drewnoakes committed:
Add initial protocol definition file

This `.proto` file defines a contract between the server `Aspire.Hosting` and client `Aspire.Dashboard`.

It supports:

- Getting basic information about the server (name and version).
- Subscribing to resources, receiving both an initial snapshot and a stream of updates as they happen.
- A heartbeat mechanism, to detect when the connection has dropped and allow prompting the user and/or reconnecting.
- Support for arbitrary kinds of resources. Not limited to known types (project, container, executable).
- Support for arbitrary commands on resources, intended for actions such as start/stop/refresh/restart/etc. The server defines the commands available for each resource, and the UI can show them.

It's added to both the server and client projects, and code generation is enabled.

drewnoakes committed:
Remove heartbeat

Turns out gRPC runs over HTTP/2 which includes a keep alive ping. We don't need to model those messages in our API thankfully.

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,166 @@
+////////////////////////////////////////////
+
+message ResourceType {
+    // Unique name for the resource type. Not for display. Equivalent to ResourceId.resource_type
```
```suggestion
    // Unique name for the resource type. Equivalent to ResourceId.resource_type
    // If "display_name" is omitted, this value will be used in UIs.
```

drewnoakes committed:
Clarify ResourceType field documentation

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,168 @@
+    repeated string values = 1;
+}
+
+message AdditionalData {
+    // TODO do we need separate display value(s)?
+    string name = 1;
+    // Optional namespace, e.g. "container", "executable", "project", ...
+    optional string namespace = 2;
+    // A single value will be most common. Also support lists, to avoid escaping.
+    oneof kind {
+        string value = 3;
+        StringArray values = 4;
+    }
+}
```
```suggestion
message AdditionalData {
    // Name of the data item, e.g. "id", "image", "path", ...
    string name = 1;
    // Optional namespace, e.g. "container", "executable", "project", ...
    optional string namespace = 2;
    // Optional display name, may be localized
    optional string display_name = 3;
    // A single value will be most common. Also support lists, to avoid escaping.
    oneof kind {
        string value = 4;
        StringArray values = 5;
    }
}
```

drewnoakes committed:
Add optional display name to additional data

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+    repeated Endpoint endpoints = 7;
+    repeated ResourceCommandRequest commands = 8;
+
+    // TODO potentially remove endpoints/env vars/anything that might not be general purpose
```
@smitpatel this was a concern you raised. Are some of the fields here too specific, and if so can they be made more general-purpose? The `additional_data` collection below might help.

daniv-msft commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+////////////////////////////////////////////
+
+message ResourceCommandRequest {
+    string command_type = 1;
```
Could you please share the list of the command_type that are in use today?

drewnoakes replied:
Currently there are none that I'm aware of. This is for future, so the question is whether the API shape is sufficient for the kinds of interactions we might want. Do you have some examples you might like to see?

daniv-msft replied:
Thanks for clarifying! I don't have examples. Actually, what I had in mind was more that as we'll need to implement that same API in the Container Apps environment, we will need to understand to what we can map this. If this is a future API, maybe it's fine to leave it as non-implemented for now on our side.

drewnoakes replied:
Each server side could decide which commands it wanted to expose, and the set could be empty. There's no requirement to use this, but it could be handy when you need it.

daniv-msft commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
```
There will likely be some discrepancies between the values populated there locally and the ones available when running in the cloud. The `endpoints` field for example, won't be as open in the Cloud as what it is locally. I'm not sure yet of what we'll be able to provide there, but it's possible that we need to have some form of "nullable endpoint" concept here to say that an endpoint isn't available from outside of the cluster.

drewnoakes replied:
Is there something similar to an endpoint you'd be interested in showing for such resources, or would the field just be omitted. Currently it's `repeated` where zero items is valid.

Another option is we move it to additional data.

Another option is we allow modelling hyperlinks for resources, similar to how we have commands. Maybe they're just a different kind of command (link vs. button).

daniv-msft replied:
It would feel OK for me to have these endpoints being displayed prominently in the local dashboard, and being shown as "Internal"/"In-cluster" once it is deployed on Container Apps. I think that it makes sense for the developers to see such endpoints locally, and to have a more restricted experience once things are deployed publicly.
If we could have a way to specify this (rather than just setting the endpoint value to "Internal", which wouldn't be a valid URL) in the API it would be great however.

drewnoakes replied:
@daniv-msft the current `Endpoint` message requires a name, but the addresses and port are optional. I think you could achieve what you wanted using the current design.

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,166 @@
+    // A single value will be most common. Also support lists, to avoid escaping.
+    oneof kind {
+        string value = 3;
+        StringArray values = 4;
```
Yes, unfortunately. No modifier (optional, required, repeated) can be on a member of a oneof.

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+
+import "google/protobuf/timestamp.proto";
+
+////////////////////////////////////////////
```
I expect these to go over this channel too. Initially we'll implement the protocol here in-proc. It may need further extension before we can cross the process/network boundary.

drewnoakes replied:
I pushed an update that includes a first draft of logs and telemetry.

daniv-msft replied:
Thanks!

drewnoakes replied:
Sounds like console logs are in scope, but telemetry will happen via standard OTLP mechanisms, so I'll remove the telemetry stuff I added.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+}
+
+enum ResourceCommandResponseKind {
+    none = 0;
```
What does "none" mean?

JamesNK replied:
none -> undefined?

JamesNK replied:
btw, enum values should be shouty case, e.g. NONE, UPDATE_RECORD, etc

drewnoakes replied:
Protobuf needs a zero-valued enum member, and it feels like making it a non-member of the enum is idiomatic, from the (internal) examples I've seen.

Undefined might be a better name. Will update. I'll change the case as well.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+// Initiates a subscription for data about resources.
+message WatchResourcesRequest {
+    // True if the client is establishing this connection because a prior one closed unexpectedly.
+    optional bool is_reconnect = 1;
```
How will the result of the request change if `is_recconect` is true? Does this imply the implementer of `DashboardService` is stateful?

drewnoakes replied:
The idea is just to allow the server to know if the client is experiencing disconnections. This would go in a log or telemetry or something like that.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+////////////////////////////////////////////
+
+message ResourceDeletion {
+    ResourceId resource_id = 1;
```
So deletion is one-at-a-time, but inserts/updates can be batched? Why the difference?

drewnoakes replied:
No, both deletions and upserts are multiples.

```
message ResourceDeletion {
    ResourceId resource_id = 1;
}

message WatchResourcesChange  {
    oneof kind {
        ResourceDeletion delete = 1;
        ResourceSnapshot upsert = 2;
    }
}

message WatchResourcesChanges {
    repeated WatchResourcesChange value = 1;
}
```

`WatchResourcesChanges` contains multiple `WatchResourcesChange`, and each of those may be either a delete or upsert.

Note that `ResourceSnapshot` is an immutable snapshot of a single resource, at a point in time. The one scenario where we send multiple resource snapshots is the initial sync, which is modelled as `WatchResourcesSnapshot`.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+    repeated EnvironmentVariable environment = 5;
+    optional int32 expected_endpoints_count = 6;
+    repeated Endpoint endpoints = 7;
+    repeated ResourceCommandRequest commands = 8;
```
I am not sure I understand the purpose of including command data in the resource data. To be able to visualize the history of commands and their outcomes?

If the command requests are part of the resource data, then why are command responses not there as well?

drewnoakes replied:
The idea is that there may be some commands that always apply to all resources of the type, so can be hosted up to that level rather than being repeated in every instance of the resource. A client would merge these with any per-instance commands to determine the set displayed to the user for a given resource.

karolz-ms replied:
OK.. so I think you are saying that `commands` are really metadata about available commands? 

If so, the suffix `Request` in the `ResourceCommandRequest` message type is not the best, because it suggests that the data this property carries is related to specific instance of command invocation, rather than command type. I would suggest renaming it to something like `ResourceCommandInfo`

drewnoakes replied:
The type is used both to define the command, and when clients send it. The expectation is that the client won't interpret the object at all, and there wouldn't be any user-editable parameters (e.g. "rename") for now at least.

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,170 @@
+message ResourceCommandRequest {
+    string command_type = 1;
+    ResourceId resource_id = 2;
+    optional string confirmation_message = 3;
```
Interesting idea. I think it could make sense. Maybe using the [`google.protobuf.Value`](https://learn.microsoft.com/en-us/aspnet/core/grpc/protobuf?view=aspnetcore-8.0#value) James pointed to earlier.

It'd be possible to add this later safely too. Do you have a specific use case in mind? That might help prioritize.

drewnoakes replied:
I added a parameter to the commands.

karolz-ms replied:
E.g. if we want to allow dashboard users to scale resources up/down, the parameter might be desired number of replicas.

drewnoakes replied:
Having user-editable parameters would be more complex, as we then need to add metadata for the UI to know how to display/constrain/validate the parameter's value. Before we ship V1 I think we'll want to revisit commands and see whether to keep parameters (and whether to keep the command support at all). It'll depend how much utility we can get out of them between now and then.

drewnoakes committed:
Review feedback

drewnoakes committed:
Merge remote-tracking branch 'upstream/main' into add-resource-server-proto

drewnoakes committed:
Add services for console and structured logs

drewnoakes committed:
Add TODO

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,260 @@
+
+message ConsoleLogLine {
+    string text = 1;
+    optional bool is_error = 2;
```
Could you add a description for `is_error` property? I have experienced people being confused between "this is a log line that describes an application error" vs "this log line was written to standard error stream" (`stderr`, descriptor 2)

drewnoakes replied:
Done.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,260 @@
+}
+
+// Maps to Microsoft.Extensions.Logging.LogLevel (in names, not numbers)
+enum StructuredLogLevel {
```
This is inconsistent with the OpenTelemetry log data model https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber and also a slippery slope, because I have seen libraries define the log levels "the other way around", namely, using lower numbers for higher severity errors. A good example is Event Tracing for Windows (ETW) https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-enabletraceex2

drewnoakes replied:
What do you suggest?

karolz-ms replied:
I would follow the OTEL log data model and be prepared that the resource service might need to do some "translation" of log levels depending on the source. 

This also is related to the discussion above https://github.com/dotnet/aspire/pull/1274/files#r1422518059 , namely the question what is the telemetry format we standardize on for the sake of populating the dashboard. Following OTEL conventions makes sense to me and makes it easier to handle differences such as 
- resource service being the store of telemetry (local dev environment) vs just exposer/transmitter of telemetry (some cloud environments, presumably)
- application services implemented in .NET vs application services implemented in other languages

drewnoakes replied:
We ended up removing these types.

karolz-ms commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,260 @@
+    int32 start_index = 2;
+    optional int32 count = 3;
+    optional StructuredLogLevel level = 4;
+    repeated StructuredLogFilter filters = 5;
```
We could also consider specifying a DSL for filters.

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -167,8 +167,93 @@ message WatchResourcesUpdate {
+
+////////////////////////////////////////////
+
+message StructuredLog {
```
Is the dashboard no longer going to host the OTLP server?

drewnoakes replied:
If there's an existing service for this, I'd like to use it. Can you provide some more info?

JamesNK replied:
Telemetry data is pushed into the dashboard with OTLP. I thought the dashboard was going to continue to host the OTLP server and store its data in memory. Has that changed?

drewnoakes replied:
Seems like there's different ideas here. @karolz-ms suggested that logs/telemetry might go over the same gRPC channel. I have been focussing mostly on the resource data flow for now, and will continue with that in the coming days. That'll give us time to discuss and work out the design here (and give me time to do a little more research).

Will the OTLP approach work in ACA to the best of your knowledge? It would be good to not implement the server side twice, once in the app host and again for ACA.

JamesNK replied:
> Seems like there's different ideas here. @karolz-ms suggested that logs/telemetry might go over the same gRPC channel.

What is the goal?

There are big implications to moving OTEL telemetry out of the dashboard. The service implementor would need to:
* Collect and store the OTEL data. They would probably need to implement an OTLP service like we've already done in the dashboard.
* Be responsible for storing the collected OTEL data in memory
* Implement the gRPC methods to so the dashboard can query the data

It wouldn't be difficult to move C# code from the dashboard project to the hosting project to fulfill this list. But a cloud provider would have a lot of work ahead of them to do this from scratch.

I wrote a doc about how OTEL works in Aspire: https://github.com/dotnet/aspire/blob/main/docs/open-telemetry-architecture.md

karolz-ms replied:
> Seems like there's different ideas here. @karolz-ms suggested that logs/telemetry might go over the same gRPC channel. 

There must have been some miscommunication--I did not ever intend to suggest such thing ☺️

davidfowl replied:
We shouldn't touch the OTLP server part of the dashboard.

drewnoakes replied:
@karolz-ms apologies, I misremembered. It was [this comment](https://github.com/dotnet/aspire/pull/1274#discussion_r1419902127) from @daniv-msft (not you) that prompted the addition of structure log support to the protocol.

Thanks all for levelling up my understanding. I've removed structured logs from the proto here and have a better idea of what we're aiming for. The console log support that @daniv-msft suggested remains.

drewnoakes committed:
Add comment

drewnoakes committed:
Merge branch 'main' into add-resource-server-proto

drewnoakes committed:
Remove obsolete TODO comment

drewnoakes committed:
Remove structured log messages

This will be handled via OTLP.

drewnoakes committed:
Updates from review meeting

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -0,0 +1,190 @@
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
```
Could `ResourceSnapshot` just be `Resource`? I think it's implicit that all the data returned from a service is a snapshot in time. Adding snapshot to the message name is unnecessary.

drewnoakes replied:
Perhaps. I try to avoid very generic names like "resource" as they tend to conflict. That said, we'll only be using these messages in two places in C#, as we'll only create/consume gRPC types at the edges, and convert to/from our own classes beyond that.

drewnoakes committed:
Split Endpoint and Service types

These have different fields.

drewnoakes committed:
Rename messages and fields

drewnoakes committed:
Add .editorconfig to solution

This allows opening the file more conveniently, and the file to participate in find-in-files.

drewnoakes commented:
/azp run
